src/arkode/arkode_mristep.c

```diff
@@ -1482,35 +1767,58 @@ int mriStep_TakeStep(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
     /* compute the implicit component */
     if (step_mem->implicit_rhs)
     {
-      retval = step_mem->fsi(ark_mem->tn, ark_mem->yn, step_mem->Fsi[0],
+      retval = step_mem->fsi(t0, ark_mem->yn, step_mem->Fsi[0],
                              ark_mem->user_data);
       step_mem->nfsi++;
       if (retval) { return ARK_RHSFUNC_FAIL; }
     }
 
+    /* Add external forcing to Fse[0] or Fsi[0], if appropriate */
+    if (step_mem->expforcing)
+    {
+      step_mem->cvals[0] = ONE;
+      step_mem->Xvecs[0] = step_mem->Fse[0];
+      nvec               = 1;
+      mriStep_ApplyForcing(step_mem, t0, ONE, &nvec);
+      N_VLinearCombination(nvec, step_mem->cvals, step_mem->Xvecs,
+                           step_mem->Fse[0]);
+    }
+    if (step_mem->impforcing)
+    {
+      step_mem->cvals[0] = ONE;
+      step_mem->Xvecs[0] = step_mem->Fsi[0];
+      nvec               = 1;
+      mriStep_ApplyForcing(step_mem, t0, ONE, &nvec);
+      N_VLinearCombination(nvec, step_mem->cvals, step_mem->Xvecs,
+                           step_mem->Fsi[0]);
+    }
```

Comment:
I think it's a bad idea to try to incorporate the ARKStep optimizations when using MRIStep as an inner stepper for telescopic MRI methods.  The main reason is because the ARKStep trick that you used results in the need to traverse all of the supplied "slow" forcing vectors every time the ARKStep RHS function(s) need to be evaluated.  This is perfectly fine, since ARKStep "lives" at the fastest time scale, and so these would need to be evaluated anyways, and their values have a natural location to be stored without requiring additional memory.

However, when MRIStep is an inner stepper, it must supply the RHS values from its own time scale as forcing vectors to an inner solver.  The only way we can avoid a combinatorial explosion of forcing vectors at the fastest scale of an n-way MRI application, is to have MRIStep evaluate and store the supplied forcing vectors at its relevant stage times.  I see two options to do this correctly:
1. Add the forcing terms to the stored `Fe` or `Fi` vectors, and preclude reuse of those RHS vectors when the forcing terms change (as I did in this PR).
2. Allocate `step_mem->stages` *additional* N_Vectors in MRIStep so that it can store the forcing function values at the stage times.

I dislike the idea of severely enlarging the memory footprint so that we can avoid a moderate number of RHS evaluations, particularly since those evaluations are assumed to be inexpensive in comparison to the slower time scale RHS functions.

As a middle-ground, I'll add two extra booleans to the MRIStep memory structure, `fe_is_current` and `fi_is_current`.  This way, in `SetForcing` I can mark only one of those for recomputation, and then throughout MRIStep I can key recomputation vs reuse for `Fse[0]` and `Fsi[0]` separately.